internal/client/consensus/grandpa/until_imported.go

```diff
@@ -0,0 +1,335 @@
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/gammazero/deque"
+)
+
+const logPendingInterval = 15 * time.Second
+
+// Something that needs to be withheld until specific blocks are available.
+//
+// For example a GRANDPA commit message which is not of any use without the corresponding block that it commits on.
+type blockUntilImported[H runtime.Hash, N runtime.Number, Blocked any] interface {
+	// Check if a new incoming item needs awaiting until a block(s) is imported.
+	NeedsWaiting(
+		input Blocked,
+		statusCheck BlockStatus[H, N],
+	) (discardWaitOrReady, error)
+
+	// called when the wait has completed. The canonical number is passed through for further checks.
+	WaitCompleted(canonNumber N) *Blocked
+}
+
+// Describes whether a given blockUntilImported (a) should be discarded, (b) is waiting for specific blocks to be
+// imported or (c) is ready to be used.
+//
+// A reason for discarding a blockUntilImported would be if a referenced block is perceived under a different number
+// than specified in the message.
+type discardWaitOrReady interface {
+	isDiscardWaitOrReady()
+}
+type discard struct{}
+
+func (discard) isDiscardWaitOrReady() {}
+
+type wait[H, N, M any] []struct {
+	TargetHash   H
+	TargetNumber N
+	Wait         M
+}
+
+func (wait[H, N, M]) isDiscardWaitOrReady() {}
+
+type ready[R any] struct {
+	Ready R
+}
+
+func (ready[R]) isDiscardWaitOrReady() {}
+
+// Buffering incoming messages until blocks with given hashes are imported.
+type untilImported[
+	H runtime.Hash,
+	N runtime.Number,
+	Header runtime.Header[N, H],
+	Blocked any,
+	M blockUntilImported[H, N, Blocked],
+] struct {
+	importNotifications <-chan api.BlockImportNotification[H, N, Header]
+	blockSyncRequester  BlockSyncRequester[H, N]
+	statusCheck         BlockStatus[H, N]
+	incomingMessages    <-chan Blocked
+	ready               deque.Deque[Blocked]
+	// Interval at which to check status of each awaited block.
+	checkPending <-chan time.Time
+	// Mapping block hashes to their block number, the point in time it was first encountered (Instant) and a list of
+	// GRANDPA messages referencing the block hash.
+	pending map[H]pendingEntry[H, N, Blocked]
+	// Queue identifier for differentiation in logs.
+	identifier string
+	// TODO: metrics
+}
+type pendingEntry[H runtime.Hash, N runtime.Number, Blocked any] struct {
+	BlockNumber N
+	LastLog     time.Time
+	Wait        []blockUntilImported[H, N, Blocked]
+}
+
+func newUntilImported[
+	H runtime.Hash,
+	N runtime.Number,
+	Header runtime.Header[N, H],
+	Blocked any,
+	M blockUntilImported[H, N, Blocked],
+](
+	importNotifications <-chan api.BlockImportNotification[H, N, Header],
+	blockSyncRequester BlockSyncRequester[H, N],
+	statusCheck BlockStatus[H, N],
+	incomingMessages <-chan Blocked,
+	identifier string,
+) untilImported[H, N, Header, Blocked, M] {
+	// how often to check if pending messages that are waiting for blocks to be imported can be checked.
```

Comment:
How often to check if pending messages can be checked? Should this be "retrieved" or "processed"?

Reply:
They call the interval `checkPending` so I think it's fine to leave the nomenclature as is.